File: apps/client/src/lib/academic/academics/entity.ts
import type { AcademicDTO } from "./dtos"

import { SexValue } from "$shared/value-objects/sex.value"
import { DateValue } from "$shared/value-objects/date.value"
import { CLf64Value } from "$shared/value-objects/cl-f64.value"
import { PlantaValue } from "./value-objects/planta.value"
import { CountryValue } from "$shared/value-objects/country.value"
import { AcademicOptionValue } from "./value-objects/option.value"

export class Academic {
	constructor(
		public id: string,
		public names: string,
		public paternalSurname: string,
		public maternalSurname: string,
		public email: string,
		public orcid: string | null,
		public sex: SexValue,
		public birthDate: DateValue,
		public joinedAt: DateValue,
		public workPosition: string | null,
		public department: string,
		public career: string | null,
		public jce: CLf64Value,
		public category: string,
		public planta: PlantaValue,
		public option: AcademicOptionValue,
		public acadCategoryHours: number | null,
		public annualDiscountHours: number,
		public nationality: CountryValue,
		public city: string,
	) {}

	public static fromDTO(dto: AcademicDTO): Academic {
		const country = CountryValue.from(dto.nationality)
		const sex = SexValue.from(dto.sex)
		const birthDate = DateValue.from(dto.birthDate)
		const joinedAt = DateValue.from(dto.joinedAt)
		const jce = CLf64Value.from(dto.jce)
		const option = AcademicOptionValue.from(dto.option)
		const planta = PlantaValue.from(dto.planta)

		return new Academic(
			dto.id,
			dto.names,
			dto.paternalSurname,
			dto.maternalSurname,
			dto.email,
			dto.orcid,
			sex,
			birthDate,
			joinedAt,
			dto.workPosition,
			dto.department,
			dto.career,
			jce,
			dto.category,
			planta,
			option,
			dto.acadCategoryHours,
			dto.annualDiscountHours,
			country,
			dto.city,
		)
	}
}

File: apps/server/src/academic/academics/entity.rs
use crate::academic::*;
use crate::shared::model_id;
use crate::university::*;

use bon::Builder;
use jiff::{Timestamp, civil::Date};
use serde::{Deserialize, Serialize};
use toasty::{Deferred, Embed, Model};

model_id! {
	struct AcademicId,
	key: "academic_id"
}

#[derive(Debug, Clone, Copy, Embed, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
#[column(rename_all = "UPPERCASE")]
pub enum Sex {
	H,
	M,
	O,
}

#[derive(Debug, Clone, Serialize, Model, Builder)]
pub struct Academic {
	#[key]
	#[builder(default = AcademicId::new())]
	pub id: AcademicId,

	#[unique]
	pub rut: String,

	#[unique]
	pub email: String,

	#[unique]
	pub orcid: Option<String>,

	pub names: String,
	pub paternal_surname: String,
	pub maternal_surname: String,
	pub sex: Sex,
	pub joined_at: Date,
	pub birth_date: Date,
	pub jce: f64,
	pub city: String,
	pub annual_discount_hours: f64,

	#[builder(default = Timestamp::now())]
	pub updated_at: Timestamp,

	#[index]
	pub nationality_code: String,

	#[index]
	pub department_id: DepartmentId,

	#[index]
	pub career_id: Option<CareerId>,

	#[index]
	pub category_option_id: AcademicCategoryOptionId,

	#[index]
	pub work_position_id: AcademicWorkPositionId,

	#[has_many]
	#[builder(default)]
	pub degrees: Deferred<Vec<Degree>>,

	#[belongs_to(key = nationality_code, references = code)]
	#[builder(default)]
	pub nationality: Deferred<Country>,

	#[belongs_to]
	#[builder(default)]
	pub department: Deferred<Department>,

	#[belongs_to]
	#[builder(default)]
	pub career: Deferred<Option<Career>>,

	#[belongs_to]
	#[builder(default)]
	pub category_option: Deferred<AcademicCategoryOption>,

	#[belongs_to]
	#[builder(default)]
	pub work_position: Deferred<AcademicWorkPosition>,
}

#[derive(Debug)]
pub struct AcademicListFilter {
	pub search: Option<String>,
	pub career_id: Option<CareerId>,
	pub department_id: Option<DepartmentId>,
	pub category_id: Option<AcademicCategoryId>,
	pub planta: Option<AcademicPlanta>,
	pub option: Option<AcademicOption>,
}

impl Academic {
	pub fn full_name(&self) -> String {
		format!(
			"{} {} {}",
			self.names, self.paternal_surname, self.maternal_surname
		)
	}
}
